src/dune_engine/action_builder.ml

```diff
@@ -309,60 +313,59 @@ struct
     end
 
     module Poly_memo = Memo.Poly.Async (Function)
-
-    let eval_pred x = Fdecl.get Dep.eval_pred x
-
     open Memo.Build.O
 
-    let rec exec : type a. a t -> (a * Dep.Set.t) Memo.Build.t = function
-      | Pure x -> Memo.Build.return (x, Dep.Set.empty)
+    let merge_facts a b = Dep.Map.union a b ~f:(fun _ a _ -> Some a)
+
+    let rec exec : type a. a t -> (a * Build_deps.fact Dep.Map.t) Memo.Build.t =
+      function
+      | Pure x -> Memo.Build.return (x, Dep.Map.empty)
       | Map (f, a) ->
         let+ a, deps_a = exec a in
         (f a, deps_a)
       | Both (a, b) ->
         let+ (a, deps_a), (b, deps_b) =
           Memo.Build.fork_and_join (fun () -> exec a) (fun () -> exec b)
         in
-        ((a, b), Dep.Set.union deps_a deps_b)
+        ((a, b), merge_facts deps_a deps_b)
       | Seq (a, b) ->
         let+ ((), deps_a), (b, deps_b) =
           Memo.Build.fork_and_join (fun () -> exec a) (fun () -> exec b)
         in
-        (b, Dep.Set.union deps_a deps_b)
+        (b, merge_facts deps_a deps_b)
       | Map2 (f, a, b) ->
         let+ (a, deps_a), (b, deps_b) =
           Memo.Build.fork_and_join (fun () -> exec a) (fun () -> exec b)
         in
-        (f a b, Dep.Set.union deps_a deps_b)
+        (f a b, merge_facts deps_a deps_b)
       | All xs ->
         let+ res = Memo.Build.parallel_map xs ~f:exec in
         let res, deps = List.split res in
-        (res, Dep.Set.union_all deps)
+        (res, List.fold_left deps ~init:Dep.Map.empty ~f:merge_facts)
       | Deps deps ->
-        let+ () = Build_deps.register_action_deps deps in
+        let+ deps = Build_deps.register_action_deps deps in
         ((), deps)
       | Paths_for_rule ps ->
-        let+ () = Build_deps.build_deps (Dep.Set.of_files_set ps) in
-        ((), Dep.Set.empty)
+        let+ _deps = Build_deps.build_deps (Dep.Set.of_files_set ps) in
```

Comment:
Seems worth adding a comment explaining why `_deps` is ignored here.

Reply:
Agreed. I introduced a function with a explicit name and a comment